processor/src/main/java/com/linecorp/decaton/processor/runtime/internal/QuotaApplierImpl.java

```diff
@@ -0,0 +1,116 @@
+import com.linecorp.decaton.processor.metrics.Metrics;
+import com.linecorp.decaton.processor.metrics.Metrics.ShapingMetrics;
+import com.linecorp.decaton.processor.runtime.PerKeyQuotaConfig.QuotaCallback;
+import com.linecorp.decaton.processor.runtime.internal.PerKeyQuotaManager.QuotaUsage;
+import com.linecorp.decaton.processor.runtime.internal.PerKeyQuotaManager.UsageType;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class QuotaApplierImpl implements QuotaApplier {
+    private final ExecutorService shapingExecutor;
+    private final Producer<byte[], byte[]> producer;
+    private final QuotaCallback callback;
+    private final ShapingMetrics metrics;
+
+    public QuotaApplierImpl(Producer<byte[], byte[]> producer,
+                QuotaCallback callback,
+                SubscriptionScope scope) {
+        this.producer = producer;
+        this.callback = callback;
+        shapingExecutor = Executors.newSingleThreadExecutor(Utils.namedThreadFactory("TaskShaper"));
+        metrics = Metrics.withTags("subscription", scope.subscriptionId()).new ShapingMetrics();
+    }
+
+    @Override
+    public boolean apply(ConsumerRecord<byte[], byte[]> record,
+                         OffsetState offsetState,
+                         QuotaUsage quotaUsage) {
+        if (quotaUsage == null || quotaUsage.type() == UsageType.COMPLY) {
+            return false;
+        }
+
+        final String topic;
+        Completion completion = offsetState.completion();
+        try {
+            topic = callback.apply(record, quotaUsage.metrics());
+        } catch (Exception e) {
+            log.error("Exception thrown from the quota callback for key: {}", Arrays.toString(record.key()), e);
+            metrics.shapingQueueingFailed.increment();
+            completion.complete();
+            return true;
+        }
+
+        ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(
+                topic, null, record.key(), record.value(), record.headers());
+        shapingExecutor.execute(() -> {
+            try {
+                producer.send(producerRecord, (r, e) -> {
+                    if (e == null) {
+                        metrics.shapingQueuedTasks.increment();
+                    } else {
+                        metrics.shapingQueueingFailed.increment();
+                        log.error("Failed to send task to the shaping topic", e);
+                    }
+                    completion.complete();
+                });
+            } catch (Exception e) {
+                log.error("Exception thrown while sending task to the shaping topic", e);
+                metrics.shapingQueueingFailed.increment();
+                completion.complete();
+            }
+        });
+        return true;
+    }
+
+    @Override
+    public void close() {
+        shapingExecutor.shutdown();
+        try {
+            shapingExecutor.awaitTermination(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
```

Comment:
hm, so this is called from `ProcessorSubscription`, and I'm bit concerned that users who set `decaton.processing.shutdown.timeout.ms` misleaded by decaton subscription actually not terminating within the timeout because of this wait.
Unlike other operations being done at `ProcessorSubscription#cleanUp`, I think this executor, which is using producer, can block at least for `max.block.ms` * number of tasks, which might in some-cases produces unexpected behavior for users?

Reply:
Hm, good point.

Not only executor termination, also KafkaProducer#close may block for a while until all previously sent records complete, which may cause confusing behavior.

Maybe we need to introduce some timeout (fixed value first, and make it configurable when there's a request?) for QuotaApplier's close procedure (without marking timeout tasks as complete).
Also we should mention it in `decaton.processing.shutdown.timeout.ms` javadoc.

I'll fix like that. Please tell me if you have any different thoughts.

Reply:
I came up with another thought.
In the first place, we don't need to wait pending shaping tasks on close at all at this point.

Instead, once the close-procedure is initiated, we should:
- Stop processing any pending shaping tasks
- Also, we should close the producer immediately to make queued batches fail

Because at this point, no more offsets will be committed so they will be re-processed by another instances anyways.